### PR TITLE
finish crud

### PR DESCRIPTION
**Pull request by hossamAmeenXware**

### CODE REVIEW

File: ecomerce/brands/models.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from django.db import models
+from django.contrib.auth.models import User
+
+class Brand(models.Model):
+    name = models.CharField(max_length=50)
+    image = models.FileField(upload_to="uploads/", max_length=100)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/models.py:7]: You can use a `related_name` parameter to use this relation by the `related_name` value.

File: ecomerce/brands/serializers.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from rest_framework import serializers
+from .models import Brand
+
+
+class BrandSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Brand
+        fields = ['id', 'name', 'image' ,'user']
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/serializers.py:8]: You must leave a new line at the end for each file.

Comment (MahmoudFarweez1) [inline ecomerce/brands/serializers.py:8]: You can use all fields without each field name.

File: ecomerce/brands/urls.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path
+from . import views
+
+urlpatterns = [
+    path('' , views.ListBrands.as_view()),
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/urls.py:5]: You can find a way one line for all URLs.

File: ecomerce/brands/views.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+from rest_framework import generics
+from .models import Brand
+from .serializers import BrandSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
+
+
+class ListBrands(generics.ListCreateAPIView):
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:8]: You can find a way to merge all these classes into one class for `CRUD`.

File: ecomerce/brands/views.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    serializer_class = BrandSerializer
+
+
+class CreateBrand(generics.CreateAPIView):
+    queryset = Brand.objects.all()
+    serializer_class = BrandSerializer
+    
+    def create(self, request, * args, ** kwargs):
+        request.data.update({'user': request.user.id})
+        serializer = self.serializer_class(data = request.data)
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:19]: Variable name must be expressed what is the variable point.

Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:19]: There no white space in `data = request.data`.

File: ecomerce/brands/views.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        if serializer.is_valid():
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:20]: You can use the serializer variable for validation.
`serializer.is_valid(raise_exception=True)`.

File: ecomerce/brands/views.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            serializer.save()
+            return Response(serializer.data, status = status.HTTP_201_CREATED)
+        else :
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:23]: If you implement the above comment, you will not need this `else` statement.

File: ecomerce/brands/views.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status = status.HTTP_200_OK)
+        else :
+            return Response(serializer.errors, status = status.HTTP_400_BAD_REQUEST)
+        
+
+class DeleteBrand(generics.DestroyAPIView):
+    queryset = Brand.objects.all()
+    serializer_class = BrandSerializer
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:49]: New line at the end of the file.

File: ecomerce/ecomerce/settings.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+# Application definition
+
+INSTALLED_APPS = [
+    'django.contrib.admin',
+    'django.contrib.auth',
+    'django.contrib.contenttypes',
+    'django.contrib.sessions',
+    'django.contrib.messages',
+    'django.contrib.staticfiles',
+    # 'products.apps.ProductsConfig',
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/settings.py:42]: Don't push commented code.

File: ecomerce/ecomerce/settings.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    'brands',
+    'modules',
+    'stocks',
+    'products',
+    'rest_framework',
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/settings.py:47]: Add the packages before the app's name.

File: ecomerce/ecomerce/settings.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    'rest_framework_simplejwt',
+]
+
+REST_FRAMEWORK = {
+    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.PageNumberPagination',
+    'PAGE_SIZE': 10 ,
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/settings.py:55]: Why is this an empty line?

File: ecomerce/ecomerce/settings.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+]
+
+REST_FRAMEWORK = {
+    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.PageNumberPagination',
+    'PAGE_SIZE': 10 ,
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+        
+        'rest_framework_simplejwt.authentication.JWTAuthentication',
+    ),
+
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/settings.py:58]: See above comment.

File: ecomerce/ecomerce/settings.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    'AUTH_TOKEN_CLASSES': ('rest_framework_simplejwt.tokens.AccessToken',),
+    'TOKEN_TYPE_CLAIM': 'token_type',
+    'TOKEN_USER_CLASS': 'rest_framework_simplejwt.models.TokenUser',
+
+    'JTI_CLAIM': 'jti',
+
+    'SLIDING_TOKEN_REFRESH_EXP_CLAIM': 'refresh_exp',
+    'SLIDING_TOKEN_LIFETIME': timedelta(minutes=5),
+    'SLIDING_TOKEN_REFRESH_LIFETIME': timedelta(days=1),
+}
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/settings.py:181]: New line at the end of the file.

File: ecomerce/modules/migrations/0002_module_user.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 3.2 on 2022-01-11 09:04
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/migrations/0002_module_user.py:1]: See above migration comment.

File: ecomerce/modules/models.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+from django.contrib.auth.models import User
+import brands
+from brands.models import Brand
+
+
+class Module(models.Model):
+    name = models.CharField(max_length=50)
+    brand = models.ForeignKey(Brand, on_delete=models.CASCADE)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+# Create your models here.
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/models.py:11]: Remove unused lines or comments.

File: ecomerce/modules/serializers.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
+from .models import Module
+
+class ModuleSerializer(serializers.ModelSerializer):
+    class Meta:
+        fields = ['id','name' , 'brand','user']
+        model = Module
+        
+
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/serializers.py:9]: One empty line at the end of the file.

File: ecomerce/modules/serializers.py [hunk 17]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
+from .models import Module
+
+class ModuleSerializer(serializers.ModelSerializer):
+    class Meta:
+        fields = ['id','name' , 'brand','user']
+        model = Module
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/serializers.py:7]: Swap this line and the line before.

File: ecomerce/modules/serializers.py [hunk 18]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
+from .models import Module
+
+class ModuleSerializer(serializers.ModelSerializer):
+    class Meta:
+        fields = ['id','name' , 'brand','user']
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/serializers.py:6]: If you will present all fields, you can replace this line.

File: ecomerce/modules/urls.py [hunk 19]
```diff
@@ -0,0 +1,11 @@
+from django.urls.conf import path
+from . import views 
+
+
+urlpatterns = [
+    path('' , views.ListModules.as_view()),
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/urls.py:6]: You can replace these lines with one line.

File: ecomerce/modules/views.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status = status.HTTP_200_OK)
+        else :
+            return Response(serializer.errors, status = status.HTTP_400_BAD_REQUEST)
+
+class DeleteModule(generics.DestroyAPIView):
+    queryset = Module.objects.all()
+    serializer_class = ModuleSerializer
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/views.py:50]: Empty line.

File: ecomerce/products/models.py [hunk 21]
```diff
@@ -0,0 +1,13 @@
+from django.db import models
+from django.contrib.auth.models import User
+from modules.models import Module
+
+
+class Product(models.Model):
+    name     = models.CharField( max_length=50)
```
Comment (MahmoudFarweez1) [inline ecomerce/products/models.py:7]: Why are these spaces?
You just need one space after the variable name.
`name = models...`

Comment (MahmoudFarweez1) [inline ecomerce/products/models.py:7]: No spaces in `()`

File: ecomerce/products/models.py [hunk 22]
```diff
@@ -0,0 +1,13 @@
...
+    quantity = models.IntegerField()
+    price = models.FloatField()
+    discount = models.FloatField()
+    module   = models.ForeignKey(Module, on_delete=models.CASCADE)
+    user     = models.ForeignKey(User,  on_delete=models.CASCADE)
+
```
Comment (MahmoudFarweez1) [inline ecomerce/products/models.py:13]: See above comments.

File: ecomerce/products/serializers.py [hunk 23]
```diff
@@ -0,0 +1,7 @@
+from rest_framework import serializers
+from .models import Product
+
+class ProductSerializer(serializers.ModelSerializer):
+    class Meta:
+        fields = "__all__"
+        model = Product
```
Comment (MahmoudFarweez1) [inline ecomerce/products/serializers.py:7]: See above comments.

Comment (MahmoudFarweez1) [inline ecomerce/products/serializers.py:7]: Swap these 2 lines.

File: ecomerce/products/urls.py [hunk 24]
```diff
@@ -0,0 +1,9 @@
+from django.urls import path
+from . import views
+urlpatterns = [
+    path('' , views.ListProducts.as_view()),
```
Comment (MahmoudFarweez1) [inline ecomerce/products/urls.py:4]: See above comments.

File: ecomerce/products/views.py [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status = status.HTTP_200_OK)
+        else :
+            return Response(serializer.errors, status = status.HTTP_400_BAD_REQUEST)
+        
+
+class DeleteProduct(generics.DestroyAPIView):
+    queryset = Product.objects.all()
+    serializer_class = ProductSerializer
```
Comment (MahmoudFarweez1) [inline ecomerce/products/views.py:53]: See above comments.

File: ecomerce/stocks/models.py [hunk 26]
```diff
@@ -0,0 +1,7 @@
+from django.db import models
+from django.contrib.auth.models import User
+
+class Stock(models.Model):
+    name=models.CharField( max_length=50)
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/models.py:5]: See above comments.

File: ecomerce/stocks/serializers.py [hunk 27]
```diff
@@ -0,0 +1,8 @@
+from rest_framework import serializers
+from .models import Stock
+
+
+class StockSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Stock
+        fields = ['id', 'name','user']
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/serializers.py:8]: See above comments.

File: ecomerce/stocks/urls.py [hunk 28]
```diff
@@ -0,0 +1,9 @@
+from . import views
+from django.urls import path
+urlpatterns = [
+    path('' , views.ListStocks.as_view()),
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/urls.py:4]: See above comments.

File: ecomerce/stocks/views.py [hunk 29]
```diff
@@ -0,0 +1,52 @@
+from rest_framework import generics
+from .models import Stock
+from .serializers import StockSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
+
+class ListStocks(generics.ListCreateAPIView):
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/views.py:7]: See above comments.

Comment (MahmoudFarweez1) [inline ecomerce/stocks/views.py:7]: You must leave 2 lines before class.

File: ecomerce/stocks/views.py [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status = status.HTTP_200_OK)
+        else :
+            return Response(serializer.errors, status = status.HTTP_400_BAD_REQUEST)
+
+
+class DeleteStock(generics.DestroyAPIView):
+    queryset = Stock.objects.all()
+    serializer_class = StockSerializer
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/views.py:52]: See above comments.

File: .gitignore [hunk 31]
```diff
@@ -0,0 +1,3 @@
+__pycache__/
+venv
+.vscode-root
```
Comment (MahmoudFarweez1) [inline .gitignore:3]: Must be an empty line at the end of the file.

File: ecomerce/brands/models.py [hunk 32]
```diff
@@ -0,0 +1,7 @@
+from django.db import models
+from django.contrib.auth.models import User
+
+class Brand(models.Model):
+    name = models.CharField(max_length=50)
+    image = models.FileField(upload_to="uploads/", max_length=100)
+    user = models.ForeignKey(User,related_name='owner',on_delete=models.CASCADE)
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/models.py:7]: Add whitespace after `,`

Comment (MahmoudFarweez1) [inline ecomerce/brands/models.py:7]: `related_name` must be referring to the table and field name like `brand_users`.

File: ecomerce/brands/serializers.py [hunk 33]
```diff
@@ -0,0 +1,8 @@
+from rest_framework import serializers
+from .models import Brand
+
+class BrandSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Brand
+        fields = "__all__"
+
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/serializers.py:8]: Remove extra empty lines.

File: ecomerce/brands/tests.py [hunk 34]
```diff
@@ -0,0 +1,15 @@
+from django.test import TestCase
+from django.contrib.auth.models import User
+from .models import Brand
+
+# Create your tests here.
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/tests.py:5]: Delete unuseful comments

File: ecomerce/brands/tests.py [hunk 35]
```diff
@@ -0,0 +1,15 @@
...
+
+
+class BrandModelTests(TestCase):
+
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/tests.py:9]: Don't need this empty line.

File: ecomerce/brands/urls.py [hunk 36]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path,include
+from . import views
+from rest_framework.routers import DefaultRouter
+
+router = DefaultRouter()
+router.register(r'', views.BrandViewSet)
+
+urlpatterns = [
+    path('', include(router.urls)),
+]
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/urls.py:10]: Empty line

File: ecomerce/brands/urls.py [hunk 37]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path,include
+from . import views
+from rest_framework.routers import DefaultRouter
+
+router = DefaultRouter()
+router.register(r'', views.BrandViewSet)
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/urls.py:6]: Must be `router.register(r'brands', views.BrandViewSet)`
Here not in the main URLs file.

File: ecomerce/brands/views.py [hunk 38]
```diff
@@ -0,0 +1,26 @@
...
+
+class BrandViewSet(viewsets.ModelViewSet):
+    queryset = Brand.objects.all()
+    serializer_class = BrandSerializer
+    
+    def create(self, request, * args, ** kwargs):
+        request.data._mutable = True
+        request.data.update({'user': request.user.id})
+        brand_serializer = self.serializer_class(data=request.data)
+        if brand_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:15]: Don't need this if statement, `raise_exception=True` will return 400 bad requests automatically.

File: ecomerce/brands/views.py [hunk 39]
```diff
@@ -0,0 +1,26 @@
...
+            brand_serializer.save()
+            return Response(brand_serializer.data, status=status.HTTP_201_CREATED)
+        
+    def update(self, request, *args, **kwargs):
+        instance = self.get_object()
+        request.data._mutable = True
+        request.data.update({'user': instance.user.id})
+        brand_serializer = self.serializer_class(instance, data=request.data)
+        if brand_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:24]: See above comments

File: ecomerce/ecomerce/urls.py [hunk 40]
```diff
@@ -0,0 +1,33 @@
...
+    TokenRefreshView,
+)
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+        ################# login ###################
+    path('api/token/', TokenObtainPairView.as_view(), name='token_obtain_pair'),
+    path('api/token/refresh/', TokenRefreshView.as_view(), name='token_refresh'),
+    
+    path('brands/'  , include('brands.urls')),
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/urls.py:29]: `path('brands/', include('brands.urls')),` remove whitespace before `,`

File: ecomerce/modules/models.py [hunk 41]
```diff
@@ -0,0 +1,8 @@
+from django.db import models
+from django.contrib.auth.models import User
+from brands.models import Brand
+
+class Module(models.Model):
+    name = models.CharField(max_length=50)
+    brand = models.ForeignKey(Brand,related_name="brand", on_delete=models.CASCADE)
+    user = models.ForeignKey(User,related_name="created_by", on_delete=models.CASCADE)
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/models.py:8]: Whitespaces

Comment (MahmoudFarweez1) [inline ecomerce/modules/models.py:8]: `related_name="module_users"`

File: ecomerce/modules/urls.py [hunk 42]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path,include
+from . import views 
+from rest_framework.routers import DefaultRouter
+
+router = DefaultRouter()
+router.register(r'', views.ModuleViewSet)
+
+urlpatterns = [
+    path('', include(router.urls)),
+]
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/urls.py:10]: Empty lines

File: ecomerce/modules/views.py [hunk 43]
```diff
@@ -0,0 +1,26 @@
...
+
+class ModuleViewSet(viewsets.ModelViewSet):
+    queryset = Module.objects.all()
+    serializer_class = ModuleSerializer
+    
+    def create(self, request, * args, ** kwargs):
+        request.data._mutable = True
+        request.data.update({'user': request.user.id})
+        module_serializer = self.serializer_class(data=request.data)
+        if module_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/views.py:15]: See above comments

File: ecomerce/modules/views.py [hunk 44]
```diff
@@ -0,0 +1,26 @@
...
+            module_serializer.save()
+            return Response(module_serializer.data, status=status.HTTP_201_CREATED)
+           
+    def update(self, request, *args, **kwargs):
+        request.data._mutable = True
+        instance = self.get_object()
+        request.data.update({'user': instance.user.id})
+        module_serializer = self.serializer_class(instance, data=request.data)
+        if module_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/views.py:24]: See above comments

File: ecomerce/products/serializers.py [hunk 45]
```diff
@@ -0,0 +1,8 @@
+from rest_framework import serializers
+from .models import Product
+
+class ProductSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Product
+        fields = "__all__"
+
```
Comment (MahmoudFarweez1) [inline ecomerce/products/serializers.py:8]: See above comments

File: ecomerce/products/urls.py [hunk 46]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path,include
+from . import views 
+from rest_framework.routers import DefaultRouter
+
+router = DefaultRouter()
+router.register(r'', views.ProductViewSet)
+
+urlpatterns = [
+    path('', include(router.urls)),
+]
```
Comment (MahmoudFarweez1) [inline ecomerce/products/urls.py:10]: See above comments

File: ecomerce/products/views.py [hunk 47]
```diff
@@ -0,0 +1,26 @@
...
+
+class ProductViewSet(viewsets.ModelViewSet):
+    queryset = Product.objects.all()
+    serializer_class = ProductSerializer
+    
+    def create(self, request, * args, ** kwargs):
+        request.data._mutable = True
+        request.data.update({'user': request.user.id})
+        product_serializer = self.serializer_class(data=request.data)
+        if product_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/products/views.py:15]: See above comments

File: ecomerce/products/views.py [hunk 48]
```diff
@@ -0,0 +1,26 @@
...
+            product_serializer.save()
+            return Response(product_serializer.data, status=status.HTTP_201_CREATED)
+        
+    def update(self, request, *args, **kwargs):
+        instance = self.get_object()
+        request.data._mutable = True
+        request.data.update({'user': instance.user.id})
+        product_serializer = self.serializer_class(instance, data=request.data)
+        if product_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/products/views.py:24]: See above comments

File: ecomerce/stocks/urls.py [hunk 49]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path,include
+from . import views 
+from rest_framework.routers import DefaultRouter
+
+router = DefaultRouter()
+router.register(r'', views.StockViewSet)
+
+urlpatterns = [
+    path('', include(router.urls)),
+]
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/urls.py:10]: See above comments

File: ecomerce/stocks/views.py [hunk 50]
```diff
@@ -0,0 +1,26 @@
...
+
+class StockViewSet(viewsets.ModelViewSet):
+    queryset = Stock.objects.all()
+    serializer_class = StockSerializer
+    
+    def create(self, request, * args, ** kwargs):
+        request.data._mutable = True
+        request.data.update({'user': request.user.id})
+        stock_serializer = self.serializer_class(data=request.data)
+        if stock_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/views.py:15]: See above comments

File: ecomerce/stocks/views.py [hunk 51]
```diff
@@ -0,0 +1,26 @@
...
+            stock_serializer.save()
+            return Response(stock_serializer.data, status=status.HTTP_201_CREATED)
+        
+    def update(self, request, *args, **kwargs):
+        instance = self.get_object()
+        request.data._mutable = True
+        request.data.update({'user': instance.user.id})
+        stock_serializer = self.serializer_class(instance, data=request.data)
+        if stock_serializer.is_valid(raise_exception=True):
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/views.py:24]: See above comments

File: ecomerce/modules/models.py (hunk 41 above)
Comment (hossamAmeenXware) [reply ecomerce/modules/models.py:8]: why "users" not "user"?

File: ecomerce/brands/views.py [hunk 52]
```diff
@@ -0,0 +1,26 @@
+from .models import Brand
+from .serializers import BrandSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:4]: space after `,`

File: ecomerce/brands/views.py [hunk 53]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class BrandViewSet(viewsets.ModelViewSet):
+    queryset = Brand.objects.all()
+    serializer_class = BrandSerializer
+    
+    def create(self, request, * args, ** kwargs):
```
Comment (MahmoudFarweez1) [inline ecomerce/brands/views.py:11]: `def create(self, request, *args, **kwargs):`

File: ecomerce/ecomerce/urls.py [hunk 54]
```diff
@@ -0,0 +1,33 @@
...
+    TokenRefreshView,
+)
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+        ################# login ###################
+    path('api/token/', TokenObtainPairView.as_view(), name='token_obtain_pair'),
+    path('api/token/refresh/', TokenRefreshView.as_view(), name='token_refresh'),
+    
+    path('', include('brands.urls')),
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/urls.py:29]: You can add `'api/'` before each URL.

File: ecomerce/ecomerce/urls.py [hunk 55]
```diff
@@ -0,0 +1,33 @@
...
+    path('', include('modules.urls') ),
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/urls.py:30]: See above comment

File: ecomerce/ecomerce/urls.py [hunk 56]
```diff
@@ -0,0 +1,33 @@
...
+    path('', include('stocks.urls')),
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/urls.py:31]: See above comment

File: ecomerce/ecomerce/urls.py [hunk 57]
```diff
@@ -0,0 +1,33 @@
...
+    path('', include('products.urls')),
```
Comment (MahmoudFarweez1) [inline ecomerce/ecomerce/urls.py:32]: See above comment

File: ecomerce/modules/models.py [hunk 58]
```diff
@@ -0,0 +1,8 @@
+from django.db import models
+from django.contrib.auth.models import User
+from brands.models import Brand
+
+class Module(models.Model):
+    name = models.CharField(max_length=50)
+    brand = models.ForeignKey(Brand, related_name="brand", on_delete=models.CASCADE)
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/models.py:7]: `related_name="module_brand"`

File: ecomerce/modules/views.py [hunk 59]
```diff
@@ -0,0 +1,26 @@
...
+from .serializers import ModuleSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
+
+
+class ModuleViewSet(viewsets.ModelViewSet):
+    queryset = Module.objects.all()
+    serializer_class = ModuleSerializer
+    
+    def create(self, request, * args, ** kwargs):
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/views.py:11]: See above comment

File: ecomerce/modules/views.py [hunk 60]
```diff
@@ -0,0 +1,26 @@
...
+        request.data._mutable = True
+        request.data.update({'user': request.user.id})
+        module_serializer = self.serializer_class(data=request.data)
+        module_serializer.is_valid(raise_exception=True)
+        module_serializer.save()
+        return Response(module_serializer.data, status=status.HTTP_201_CREATED)
+           
+    def update(self, request, *args, **kwargs):
+        request.data._mutable = True
```
Comment (MahmoudFarweez1) [inline ecomerce/modules/views.py:20]: Replace this line with the next line.

File: ecomerce/products/models.py [hunk 61]
```diff
@@ -0,0 +1,13 @@
+from django.db import models
+from django.contrib.auth.models import User
+from modules.models import Module
+
+
+class Product(models.Model):
+    name     = models.CharField(max_length=50)
+    quantity = models.IntegerField()
+    price    = models.FloatField()
```
Comment (MahmoudFarweez1) [inline ecomerce/products/models.py:9]: See above comment

File: ecomerce/products/models.py [hunk 62]
```diff
@@ -0,0 +1,13 @@
...
+    discount = models.FloatField()
+    module   = models.ForeignKey(Module,related_name="module", on_delete=models.CASCADE)
```
Comment (MahmoudFarweez1) [inline ecomerce/products/models.py:11]: See above comments

Comment (MahmoudFarweez1) [inline ecomerce/products/models.py:11]: Spacesssssssssssssssssssssssssssssssssssssssss

Comment (MahmoudFarweez1) [inline ecomerce/products/models.py:11]: `related_name="product_module`

File: ecomerce/products/urls.py [hunk 63]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path,include
+from . import views 
+from rest_framework.routers import DefaultRouter
+
+router = DefaultRouter()
+router.register(r'', views.ProductViewSet)
```
Comment (MahmoudFarweez1) [inline ecomerce/products/urls.py:6]: See above comments

File: ecomerce/products/views.py [hunk 64]
```diff
@@ -0,0 +1,26 @@
...
+from .serializers import ProductSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
+
+
+class ProductViewSet(viewsets.ModelViewSet):
+    queryset = Product.objects.all()
+    serializer_class = ProductSerializer
+    
+    def create(self, request, * args, ** kwargs):
```
Comment (MahmoudFarweez1) [inline ecomerce/products/views.py:11]: See above comments

File: ecomerce/products/views.py [hunk 65]
```diff
@@ -0,0 +1,26 @@
+from .models import Product
+from .serializers import ProductSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
```
Comment (MahmoudFarweez1) [inline ecomerce/products/views.py:4]: See above comment

File: ecomerce/stocks/models.py [hunk 66]
```diff
@@ -0,0 +1,6 @@
+from django.db import models
+from django.contrib.auth.models import User
+
+class Stock(models.Model):
+    name=models.CharField(max_length=50)
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/models.py:5]: See above comment

File: ecomerce/stocks/models.py [hunk 67]
```diff
@@ -0,0 +1,6 @@
...
+    user = models.ForeignKey(User,related_name="owner_stock", on_delete=models.CASCADE)
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/models.py:6]: See above comments

File: ecomerce/stocks/test_apiView.py [hunk 68]
```diff
@@ -0,0 +1,16 @@
+from django.test import TestCase
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/test_apiView.py:1]: Why this file?

File: ecomerce/stocks/urls.py [hunk 69]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path,include
+from . import views 
+from rest_framework.routers import DefaultRouter
+
+router = DefaultRouter()
+router.register(r'', views.StockViewSet)
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/urls.py:6]: See above comments

File: ecomerce/stocks/views.py [hunk 70]
```diff
@@ -0,0 +1,26 @@
...
+from .serializers import StockSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
+
+
+class StockViewSet(viewsets.ModelViewSet):
+    queryset = Stock.objects.all()
+    serializer_class = StockSerializer
+    
+    def create(self, request, * args, ** kwargs):
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/views.py:11]: See above comment

File: ecomerce/stocks/views.py [hunk 71]
```diff
@@ -0,0 +1,26 @@
+from .models import Stock
+from .serializers import StockSerializer
+from rest_framework.response import Response
+from rest_framework import status,viewsets
```
Comment (MahmoudFarweez1) [inline ecomerce/stocks/views.py:4]: See above comment

**Review by MahmoudFarweez1 — APPROVED** [review]

**Pull request merged**